pgjdbc/src/main/java/org/postgresql/util/LazyCleaner.java

```diff
@@ -0,0 +1,196 @@
+    if (alive == (keepAliveCleanable != null)) {
+      return this;
+    }
+
+    if (alive) {
+      keepAliveCleanable = register(this, null);
+    } else {
+      if (keepAliveCleanable != null) {
+        keepAliveCleanable.clean();
+      }
+      keepAliveCleanable = null;
+    }
+    return this;
+  }
+
+  public Cleanable register(Object obj, @Nullable CleaningAction action) {
+    return add(new Node(obj, action));
+  }
+
+  public int getWatchedCount() {
+    return watchedCount;
+  }
+
+  public boolean isThreadRunning() {
+    return threadRunning;
+  }
+
+  private synchronized boolean checkEmpty() {
+    if (first == null) {
+      threadRunning = false;
+      return true;
+    }
+    return false;
+  }
+
+  private synchronized Node add(Node node) {
+    if (first != null) {
+      node.next = first;
+      first.prev = node;
+    }
+    first = node;
+    watchedCount++;
+
+    if (!threadRunning) {
+      threadRunning = startThread();
+    }
+    return node;
+  }
+
+  private boolean startThread() {
+    Thread thread = threadFactory.newThread(new Runnable() {
+      public void run() {
+        while (true) {
+          try {
+            Node ref = (Node) queue.remove(threadTtl);
+            if (ref != null) {
+              ref.onClean(true);
+            } else if (LazyCleaner.this.checkEmpty()) {
+              break;
+            }
+          } catch (Throwable e) {
+            // Ignore exceptions from the cleanup action (including interruption of cleanup thread)
+            LOGGER.log(Level.WARNING, "Unexpected exception in cleaner thread main loop", e);
+          }
+        }
+      }
+    });
+    if (thread != null) {
+      thread.start();
+      return true;
+    }
+    return false;
+  }
+
+  private synchronized boolean remove(Node node) {
+    // If already removed, do nothing
+    if (node.next == node) {
+      return false;
+    }
+
+    // Update list
+    if (first == node) {
+      first = node.next;
+    }
+    if (node.next != null) {
+      node.next.prev = node.prev;
+    }
+    if (node.prev != null) {
+      node.prev.next = node.next;
+    }
+
+    // Indicate removal by pointing the cleaner to itself
+    node.next = node;
+    node.prev = node;
+
+    watchedCount--;
+    return true;
+  }
+
+  private class Node extends PhantomReference<Object> implements Cleanable, CleaningAction {
+    private final @Nullable CleaningAction action;
+    private @Nullable Node prev = null;
+    private @Nullable Node next = null;
+
+    Node(Object referent, @Nullable CleaningAction action) {
+      super(referent, queue);
+      this.action = action;
+      //Objects.requireNonNull(referent); // poor man`s reachabilityFence
+    }
+
+    public void clean() {
+      onClean(false);
+    }
+
+    public void onClean(boolean leak) {
+      if (!remove(this)) {
+        return;
+      }
+      try {
+        if (action != null) {
+          action.onClean(leak);
+        }
```

Comment:
If we skip the cleaning action on null, then there was no purpose of queueing the action in the first place